### PR TITLE
Lisa dev merge

### PR DESCRIPTION
**Pull request by ConWea**
Added options for plugin waveforms that compute their own arm response within the fd_seqeuence framework. Hard coded options for LISA detector within relbin.py for inference. Next step is to remove the LISA detector flags and generalise to all waveforms that compute their own arm response.

### CODE REVIEW

File: examples/lisa/run_inference.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
```
Comment (ahnitz) [inline examples/lisa/run_inference.sh:1]: Add script to download the data files for the test example. Don't commit large data files though.

File: pycbc/waveform/waveform.py [hunk 2]
```diff
@@ -490,9 +491,10 @@ def _lalsim_fd_sequence(**p):
 for apx in _lalsim_enum:
     fd_sequence[apx] = _lalsim_fd_sequence
 
+
```
Comment (ahnitz) [inline pycbc/waveform/waveform.py:494]: add function to get list of sequency and det_sequence approximants

File: pycbc/inference/models/relbin.py [hunk 3]
```diff
@@ -169,10 +169,14 @@ def __init__(
             variable_params, data, low_frequency_cutoff, **kwargs
         )
 
+        # FIXME: This should be set properly in some way
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:172]: Replace with check on which interface provides the approximant

File: pycbc/inference/models/relbin.py [hunk 4]
```diff
@@ -440,20 +481,26 @@ def _loglr(self):
...
+                dt = 0
+                dtc = -end_time
+            else:
+                det = self.det[ifo]
+                fp, fc = det.antenna_pattern(p["ra"], p["dec"],
+                                             p["polarization"], times)
+                dt = det.time_delay_from_earth_center(p["ra"], p["dec"], times)
+                dtc = p["tc"] + dt - end_time
 
+            hp, hc = wfs[ifo]
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:503]: Directly add new cython function and set set.lik to that if the detector response is already applied

File: pycbc/inference/models/relbin.py (hunk 3 above)
Comment (ConWea) [reply pycbc/inference/models/relbin.py]: Changed all "is_lisa" flags to check the approximant name in the list fd_det_sequence instead

File: pycbc/inference/models/relbin.py (hunk 4 above)
Comment (ConWea) [reply pycbc/inference/models/relbin.py]: Added new cython function for fd_det_seqeunce waveforms. Changed relbin.py path to use new function.

File: pycbc/inference/models/relbin.py [hunk 5]
```diff
@@ -169,10 +171,17 @@ def __init__(
             variable_params, data, low_frequency_cutoff, **kwargs
         )
 
+        # If the waveform handles the detector response internally, set
+        # self.det_response = True
+        self.det_response = False
+        if self.static_params['approximant'] in fd_det_sequence:
+            self.det_response = True
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:178]: The variable naming seems backwards to me as it could be read that the likelihood *will* handle the detector response if this is set.

File: pycbc/inference/models/relbin_cpu.pyx [hunk 6]
```diff
@@ -107,6 +107,38 @@ cpdef likelihood_parts(double [::1] freqs,
         x0 = x0n
     return hd, hh
 
+cpdef likelihood_parts_det(double [::1] freqs,
+                     double fp,
```
Comment (ahnitz) [inline pycbc/inference/models/relbin_cpu.pyx:111]: You can get rid of fp

File: pycbc/inference/models/relbin.py [hunk 7]
```diff
@@ -440,26 +489,37 @@ def _loglr(self):
...
+            hp, hc = wfs[ifo]
+            if self.det_response:
+                fp = 1
+                dtc = -end_time
+                hp, hc = wfs[ifo]
+                hdp, hhp = self.lik(freqs, fp, dtc,
+                                    hp, hc, h00,
+                                    sdat['a0'], sdat['a1'],
+                                    sdat['b0'], sdat['b1'])
+                self._current_wf_parts[ifo] = (fp, dtc, hp, hc, h00)
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py]: You shouldn't need this line.

File: pycbc/inference/models/relbin.py [hunk 8]
```diff
@@ -440,26 +489,37 @@ def _loglr(self):
...
-                                hp, hc, h00,
-                                sdat['a0'], sdat['a1'],
-                                sdat['b0'], sdat['b1'])
-
-            self._current_wf_parts[ifo] = (fp, fc, dtc, hp, hc, h00)
+            hp, hc = wfs[ifo]
+            if self.det_response:
+                fp = 1
+                dtc = -end_time
+                hp, hc = wfs[ifo]
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py]: Don't fake calling this hp, hc

File: pycbc/inference/models/relbin.py [hunk 9]
```diff
@@ -440,26 +484,38 @@ def _loglr(self):
...
+            if self.det_response:
+                fp = 1
+                dtc = -end_time
+                print(wfs[ifo])
+                hp = wfs[ifo]
+                hdp, hhp = self.lik(freqs, fp, dtc,
+                                    hp, h00,
+                                    sdat['a0'], sdat['a1'],
+                                    sdat['b0'], sdat['b1'])
+                self._current_wf_parts[ifo] = (fp, dtc, hp, h00)
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:505]: Remove this line. It is needed for multi-signal support, but we can't handle that case at the moment when using this kind of model.

File: pycbc/inference/models/relbin.py [hunk 10]
```diff
@@ -440,26 +484,38 @@ def _loglr(self):
...
-
-            self._current_wf_parts[ifo] = (fp, fc, dtc, hp, hc, h00)
+            # project waveform to detector frame if waveform does not deal
+            # with detector response. Otherwise, skip detector response.
+
+            if self.det_response:
+                fp = 1
+                dtc = -end_time
+                print(wfs[ifo])
+                hp = wfs[ifo]
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:500]: Rename from hp to something else

File: pycbc/inference/models/relbin.py [hunk 11]
```diff
@@ -340,13 +375,22 @@ def summary_product(self, h1, h2, bins, ifo):
...
-            hp, hc = get_fd_waveform_sequence(sample_points=edge, **params)
-            hp = hp.numpy()
-            hc = hc.numpy()
-            wfs.append((hp, hc))
-        return {ifo: wfs[self.ifo_map[ifo]] for ifo in self.data}
+        if self.det_response:
+            wfs = {}
+            for ifo in self.data:
+                wfs = wfs | get_fd_det_waveform_sequence(ifo,
+                                            sample_points=self.edge_unique[0],
```
[the remark this answers is not in the document]
Comment (ahnitz) [reply pycbc/inference/models/relbin.py:382]: wfs = get_fd_det_waveform_sequence(ifos=self.data.keys(), ....)

Then you don't need a loop or ifo.

[the remark this answers is not in the document]
Comment (ConWea) [reply pycbc/inference/models/relbin.py:382]: This has been fixed with adding sample_points=self.fedges[ifo] and looping over (could be the case that fedges is not the same across all data streams, could be a point of optimization in the future as this is generating AET 3 times.)

File: pycbc/inference/models/relbin.py [hunk 12]
```diff
@@ -440,26 +484,38 @@ def _loglr(self):
...
-                                hp, hc, h00,
-                                sdat['a0'], sdat['a1'],
-                                sdat['b0'], sdat['b1'])
-
-            self._current_wf_parts[ifo] = (fp, fc, dtc, hp, hc, h00)
+            # project waveform to detector frame if waveform does not deal
+            # with detector response. Otherwise, skip detector response.
+
+            if self.det_response:
+                fp = 1
```
[the remark this answers is not in the document]
Comment (ConWea) [reply pycbc/inference/models/relbin.py:497]: Removed

File: pycbc/inference/models/relbin.py (hunk 9 above)
Comment (ConWea) [reply pycbc/inference/models/relbin.py:505]: Removed

File: pycbc/inference/models/relbin.py (hunk 7 above)
Comment (ConWea) [reply pycbc/inference/models/relbin.py]: Removed

File: pycbc/inference/models/relbin_cpu.pyx (hunk 6 above)
Comment (ConWea) [reply pycbc/inference/models/relbin_cpu.pyx:111]: Removed

File: pycbc/waveform/waveform.py [hunk 13]
```diff
@@ -524,10 +526,43 @@ def get_fd_waveform_sequence(template=None, **kwds):
...
+    if input_params['approximant'] not in fd_det_sequence:
+        raise ValueError("Approximant %s not available" %
+                            (input_params['approximant']))
+    wav_gen = fd_det_sequence[input_params['approximant']]
+    if hasattr(wav_gen, 'required'):
+        required = wav_gen.required
+    else:
+        required = parameters.fd_required
+    check_args(input_params, required)
+    return wav_gen(**input_params)
```
[the remark this answers is not in the document]
Comment (ConWea) [reply pycbc/waveform/waveform.py]: Implemented now

File: pycbc/inference/models/relbin.py [hunk 14]
```diff
@@ -181,6 +190,10 @@ def __init__(
         # store fiducial waveform params
         self.fid_params = self.static_params.copy()
         self.fid_params.update(fiducial_params)
+
+        if self.no_det_response:
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:194]: Why is this part needed?

File: pycbc/inference/models/relbin.py [hunk 15]
```diff
@@ -204,12 +217,22 @@ def __init__(
             # prune low frequency samples to avoid waveform errors
             fpoints = Array(self.f[ifo].astype(numpy.float64))
             fpoints = fpoints[self.kmin[ifo]:self.kmax[ifo]+1]
-            fid_hp, fid_hc = get_fd_waveform_sequence(sample_points=fpoints,
-                                                      **self.fid_params)
+
+            if self.no_det_response:
+                if ifo not in self.no_det_waves:
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:222]: 'no_det_waves' is never used, so is this needed?

File: pycbc/inference/models/relbin.py [hunk 16]
```diff
@@ -435,34 +479,42 @@ def _loglr(self):
         p.update(self.static_params)
         wfs = self.get_waveforms(p)
 
-        hh = 0.0
-        hd = 0j
+        norm = 0.0
+        filter = 0j
```
Comment (ahnitz) [inline pycbc/inference/models/relbin.py:483]: 'filter' is a reserved name by python, so we try to avoid overloading it new code.
